tubesync/sync/tasks.py

```diff
@@ -254,6 +255,11 @@ def download_media_metadata(media_id):
             log.warn(f'Media: {source} / {media} is older than cap age '
                      f'{max_cap_age}, skipping')
             media.skip = True
+    # If the source has a search filter, check the video title matches the filter
+    if not re.search(source.filter_text,media.title):
```

Comment:
Should this be an opt-in filter (only download media that match this regex) or an opt-out filter (skip media that match this regex)?

Reply:
I thought opt-in was more intuitive for a user. Currently the filter_text is specified, and then if anything does not match the filter, it is marked as a "skip". Happy to take your suggestion to this though as once I re-write this to use an is_regex_match(media_item_title) method it'll be easy to change later. Certainly regex supports inverse matching though it's clunky. The other way would be to specify both an "include" and an "exclude" string. I.e. include "foo" but exclude "bar" but it feels unnecessarily complex.

'''
foo -> match
bar -> not matched
foo bar -> not matched
'''

Reply:
I would imagine the most common use-case for this is going to be "I want to ignore specific videos on a channel". While you can do this with positive matching regexes having to use lots of `(?!string$)`'s everywhere isn't going to be that friendly.

Personally, I think this might be better as a `exclude_regex` field. What do you think?

Reply:
I'm not sure I follow. Negative lookaheads get this behaviour you describe. Inverting it is no work though I want to check that re supports negative lookaheads as I might need to swap it for re2. My primary use case is to do both (include the word "foo" but not include the word "bar") as above.

```
Regex string: ^(?!.*foo).*$

foo
bar <<< Match
foo bar

Regex string: foo

foo <<< Match
bar
foo bar <<< Match
```

Reply:
Why not add an option to specify what the filter does? Include/Exclude.

For various use cases having an allow-list vs an ignore-list might be very powerful.

Imagine a channel with thousands of videos, and perhaps a reoccuring _title_ that appears throughout the channel, like "office update 15". It would be far easier to just set an include list of `^office update \d+$`, imo.

Or perhaps a channel that is otherwise about some topic, but has music videos strewn about it.. You might set an exclude list of `\(music\svideo\)`.

Writing regex is hard for some users and negations are tricky to reason about. Avoiding the _need_ for them would be advantageous imo. Requiring one extra switch isn't a lot of work.

Reply:
I like this idea. Essentially just a dropdown "include/exclude" though so it's clear what behaviour is currently selected (tickboxes being a bit ambiguous. I'll have a look later today or tomorrow at do a new branch to implement this.